File: AboutLayout_ObjC/AboutLayout/ArrowView.m
//
//  ArrowView.m
//  AboutLayout
//
//  Created by NixonShih on 2017/1/19.
//  Copyright © 2017年 Nixon. All rights reserved.
//

#import "ArrowView.h"

@implementation ArrowView {
    CAShapeLayer *arrowLayer;
}

- (instancetype)init {
    self = [super init];
    [self drawArrow];
    
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    
    self = [super initWithCoder:aDecoder];
    [self drawArrow];
    
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    [self drawArrow];
    
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    // 每次 layoutSubviews 的時候都重會一次線。
    arrowLayer.path = [self arrowPath].CGPath;
}

- (void)drawArrow {

    arrowLayer = [CAShapeLayer new];
    arrowLayer.frame = self.bounds;
    arrowLayer.path = [self arrowPath].CGPath;
    arrowLayer.fillColor = [UIColor whiteColor].CGColor;
    
    [self.layer addSublayer:arrowLayer];
}

- (UIBezierPath*)arrowPath {
    CGSize viewSize = self.bounds.size;
    UIBezierPath *path = [UIBezierPath new];
    [path moveToPoint:CGPointMake(0, 0)];
    [path addLineToPoint:CGPointMake(viewSize.width / 2, viewSize.height)];
    [path addLineToPoint:CGPointMake(viewSize.width, 0)];
    [path closePath];
    return path;
}

@end

File: AboutLayout_Swift/AboutLayout/ArrowView.swift
//
//  ArrowView.swift
//  AboutLayout
//
//  Created by NixonShih on 2017/1/20.
//  Copyright © 2017年 Nixon. All rights reserved.
//

import UIKit

@IBDesignable
/** 畫顆朝下的箭頭啦～！！！！！ */
class ArrowView: UIView {

    fileprivate var arrowLayer: CAShapeLayer?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        drawArrow()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        drawArrow()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        // 每次 layoutSubviews 的時候都重會一次線。
        arrowLayer?.path = arrowPath().cgPath
    }
    
    fileprivate func drawArrow() {
        
        arrowLayer = CAShapeLayer()
        
        guard let theArrowLayer = arrowLayer else { return }
        
        theArrowLayer.frame = bounds
        theArrowLayer.path = arrowPath().cgPath
        theArrowLayer.fillColor = UIColor.white.cgColor

        layer.addSublayer(theArrowLayer)
    }
    
    fileprivate func arrowPath() -> UIBezierPath {
        let viewSize = bounds.size
        let path = UIBezierPath()
        path.move(to: CGPoint(x: 0, y: 0))
        path.addLine(to: CGPoint(x: viewSize.width / 2, y: viewSize.height))
        path.addLine(to: CGPoint(x: viewSize.width, y: 0))
        path.close()
        return path
    }

}
